### PR TITLE
Apply --depth and --compact flags to --tasks-json and --tasks

### PR DESCRIPTION
**Pull request by sttk**
This is a PR about the issue #57

### CODE REVIEW

**Comment by sttk** [comment]
npm install EPERM errors frequently occured on appveyor.
It seems that this error can be fixed by using npm3.
So I modified appveyor.yml to install latest npm.

npm/npm#9696

**Comment by sttk** [comment]
Sorry. I forgot to rename the second argument of `logTasks`.

**Comment by sttk** [comment]
I've reflected the change of PR #87 for undertaker v1.0.0 and added `--no-sort` flag.

**Comment by phated** [comment]
This will need to be rebased on top of the merged undertaker 1.0 fixes.

**Comment by sttk** [comment]
I've rebased this on the current master branch.

**Comment by phated** [comment]
@sttk This will probably need many changes to support only through the config files.

**Comment by sttk** [comment]
@phated O.K. I'm going to address this from now.

**Comment by sttk** [comment]
@phated I've finished reflecting the latest master branch.

**Comment by phated** [comment]
@sttk thanks for rebasing.  However, I don't think we should expose the `--compact`, `--depth` or `--no-sort` options as command line flags.  Instead, could you implement this to only be configurable from `.gulp.*` files?

**Comment by phated** [comment]
Sorry, I forgot that we shipped `--depth` but it looks like `--compact` and `--no-sort` are being added.  I'm not sure I want to keep expanding the flags but maybe those are a must-have for @segrey's use case?

**Comment by sttk** [comment]
In his use case, gulp would be called from a program. So I think `--compact` flag is needed rather than config.

About `--depth` flag, I suppose users want to use both flag and config. For example, they usually specify depth=3 with config, and sometimes show full depth  tree with the flag.

About `--no-sort` flag, I think it may be good to be config only.

**Comment by sttk** [comment]
@phated Could you define config keys of compact, depth, sort/no-sort?

**Comment by phated** [comment]
@sttk if you are going to support `--compact` and `--depth` as flags, they should just be `flags.compact` and `flags.depth`; however, do we want to make them more descriptive (e.g. `--compact-tasks` and `--tasks-depth`)?  Do we really need the `sort`/`no-sort` option? I'm not sure who would want to turn sorting off.

**Comment by sttk** [comment]
I also think `--compact-tasks` and `--tasks-depth` are better. These names make more clear what the flags is for.

`no-sort` was requested by @erikkemperman in the issue #50.

**Comment by erikkemperman** [comment]
Yeah, and actually I might even argue the default should be to respect the order of tasks in my gulpfile, where I tend to try and find some sensible structure and keep related tasks near one another. Then I'd like to see the same order when listing the tasks. But maybe that's just me?

**Comment by phated** [comment]
@erikkemperman I think that's a sane default, which I believe is equivalent to "no sort" because I just iterate over the object (it's not guaranteed to be in insert order, but highly likely).

**Comment by erikkemperman** [comment]
Maybe I misunderstand, I thought the proposal here is to do alphabetical sorting by default, and have a "no-sort" option to get file-order. I would prefer the other way around. But again, maybe that's just me and anyway it's not a big deal :-)

**Comment by phated** [comment]
@erikkemperman the original implementation sorted by default.  This PR is adding the option to "disable" that but I'm suggesting we invert that assumption (remove sorting by default and add a `--sort-tasks` flag).

**Comment by erikkemperman** [comment]
Excellent, that's what I would expect :-)

**Comment by phated** [comment]
@sttk are you still planning to update this PR based on the discussion?

**Comment by sttk** [comment]
@phated Sorry that I'm late to update.

I implemented and I'm testing about this PR from now.

I'm changing the flag names to: `--compact-tasks`, `--tasks-depth`, `--sort-tasks`, and adding the configs: `flags.compactTasks`, `flags.tasksDepth`, `flags.sortTasks` in accordance with the flag names.

Please let me know if there are flags or configs that should be renamed or removed,

**Comment by phated** [comment]
@sttk I think all of those flags & configs are good.

**Comment by sttk** [comment]
@phated I've rebased and updated.

File: docs/CLI.md [hunk 1]
```diff
@@ -59,6 +59,15 @@ gulp has very few flags to know about. All other flags are for tasks to use if n
 **--tasks-json** [path]
     Print the task dependency tree, in JSON format, for the loaded gulpfile. The [path] argument is optional, and if given writes the JSON to the path.
 
+**--depth** [number]
+    Specify the depth of the task dependency tree to print. This flag can be used with --tasks or --tasks-json.
```
Comment (tkellen) [inline docs/CLI.md]: This should be tasks-depth, right?

File: docs/CLI.md [hunk 2]
```diff
@@ -59,6 +59,15 @@ gulp has very few flags to know about. All other flags are for tasks to use if n
...
+
+**--compact**
```
Comment (tkellen) [inline docs/CLI.md]: This should be compact-tasks, right?

**Comment by sttk** [comment]
@tkellen Thanks! I've modified what you pointed out.

File: lib/shared/cliOptions.js [hunk 3]
```diff
@@ -47,12 +47,6 @@ module.exports = {
     desc: chalk.gray(
       'Print the task dependency tree for the loaded gulpfile.'),
   },
-  depth: {
```
Comment (phated) [inline lib/shared/cliOptions.js]: This is a breaking change. Can we leave it in as an alias? Does yargs support aliases in the definition?

Comment (phated) [inline lib/shared/cliOptions.js]: If we alias this, it'll need to be documented and noted as deprecated.

File: lib/versioned/^3.7.0/index.js [hunk 4]
```diff
@@ -34,20 +37,56 @@ function execute(opts, env, config) {
...
         tree.label = config.description;
       } else {
         tree.label = 'Tasks for ' + chalk.magenta(tildify(env.configPath));
       }
-      return logTasks(tree, opts.depth, function(task) {
+      return logTasks(tree, opts, function(task) {
         return gulpInst.tasks[task].fn;
       });
     }
+    if (opts.tasksJson) {
```
Comment (phated) [inline lib/versioned/^3.7.0/index.js]: Can this addition be moved to a separate commit/PR?  I'm glad we are adding it but I want it to show up in the changelog properly.

File: test/expected/flags-tasks-depth4.txt [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+│   └── func2
+├─┬ taskB
+│ └─┬ <parallel>
+│   ├── func3
+│   └─┬ taskC
+│     └── <series>
+└─┬ default
+  └─┬ <parallel>
+    ├─┬ taskC
+    │ └── <series>
```
Comment (phated) [inline test/expected/flags-tasks-depth4.txt]: Is this considered part of the depth?

File: lib/shared/cliOptions.js (hunk 3 above)
Comment (sttk) [inline lib/shared/cliOptions.js]: I rethought about flag names that `--depth`, `--compact` and `--sort` are better than `--tasks-depth`, `--compact-tasks` and `--sort-tasks`. When we use these flags, we type a command like `gulp --tasks --tasks-depth`. The word "tasks" is redundant for users, I think.


I think good about the config names, but it is possible to group these flags like `flags.tasks.depth`, `flags.tasks.sort`, `flags.tasks.compact` (in .gulp.json, `{ flags: { tasks: { depth: 3, sort: true, compact: true } } }`).

What do you think?

File: lib/versioned/^3.7.0/index.js (hunk 4 above)
Comment (sttk) [inline lib/versioned/^3.7.0/index.js]: I got it. I'll remove this part from this PR.

File: test/expected/flags-tasks-depth4.txt (hunk 5 above)
Comment (sttk) [inline test/expected/flags-tasks-depth4.txt]: Yes. `depth` flag cuts deeper tasks forcely.

File: lib/shared/cliOptions.js (hunk 3 above)
Comment (phated) [inline lib/shared/cliOptions.js]: I'm fine with being verbose on the flag names but I think we just need to alias `depth` because we've previously defined it and removing it would be a breaking change (and I'd have to publish as 2.0.0)

File: test/expected/flags-tasks-depth4.txt (hunk 5 above)
Comment (phated) [inline test/expected/flags-tasks-depth4.txt]: Oh, okay

File: lib/shared/cliOptions.js (hunk 3 above)
Comment (sttk) [inline lib/shared/cliOptions.js]: I got it. I'll add the alias `depth`.

**Comment by sttk** [comment]
I've rebased and modified.

**Comment by phated** [comment]
Looks good, great work again @sttk 

I'm going to squash this, feel free to submit the `--tasks-json` feature for V3 afterwards.

**Comment by sttk** [comment]
@phated Thank you for merging.

**Comment by phated** [comment]
@sttk No problem.  I'm noticing some timeout problems (for a couple tests) on Appveyor.  Did you run into that at all?

**Comment by sttk** [comment]
@phated I sometimes encounter timeout problems on Appveyor, too. And in that case, I try test 2 or 3 times. I don't know this workaround.
